src/suncalc.cpp

```diff
@@ -475,61 +475,62 @@ double SolarNoonTimeSpec::computeTime(double rise, double set) const
 
 void SolarTimeSpec::getDay(const tm &current, int &mday, int &mon, int &year, int &wdays) const
 {
-	TimeSpec::getDay(current, mday, mon, year, wdays);
-
-	DateTime currentDate(&current);
-	currentDate.setYear(year);
-	currentDate.setMonth(mon);
-	currentDate.setDay(mday);
-	int min, hour;
-	getTime(currentDate.getDay(), currentDate.getMonth(), currentDate.getYear(), min, hour);
-	DateTime solarTime(currentDate);
-	solarTime.setHour(hour);
-	solarTime.setMinute(min);
-	if (solarTime < currentDate)
-	{
-		// Solar time has already been reached for the current day. Let's move
-		// to the next day that complies with constraints.
-		solarTime.tryIncreaseClosestGreaterFreeField(DateTime::Day);
-	}
-	mday = solarTime.getDay();
-	mon = solarTime.getMonth();
-	year = solarTime.getYear();
+    TimeSpec::getDay(current, mday, mon, year, wdays);
+
+    DateTime currentDate(&current);
+    currentDate.setYear(year);
+    currentDate.setMonth(mon);
+    currentDate.setDay(mday);
+    int min, hour;
+    getTime(currentDate.getDay(), currentDate.getMonth(), currentDate.getYear(), min, hour);
+    DateTime solarTime(currentDate);
+    solarTime.setHour(hour);
+    solarTime.setMinute(min);
+    if (solarTime < currentDate)
+    {
+        // Solar time has already been reached for the current day. Let's move
+        // to the next day that complies with constraints.
+        solarTime.tryIncreaseClosestGreaterFreeField(DateTime::Day);
+        solarTime.tryResolve(currentDate, DateTime::Year, DateTime::Day);
```

Comment:
This line is the actual fix. It makes sure solarTime is modified to correspond to an existing day in the calendar. So for the example reported in #47, solarTime is changed from March 32nd to April 1st. The thing is that SolarTimeSpec returns a value for day and month and PeriodicTask::findNext will then consider those values as frozen. The simplest solution is to make sure that SolarTimeSpec only returns resolved (hence projected on the actual calendar) values. 
I am not sure line 493 is even required now. And this could be executed even when solarTime >= currentDate (that would simply have no effect).